### ISSUE TITLE
Scalafix's dependencies have conflicting class files

### ISSUE THREAD

**Issue by cvogt** [scalacenter/scalafix#100]
Scalafix has these dependencies among others:

```
  ch.epfl.scala:scalafix-core_2.11:0.3.2
    com.lihaoyi:sourcecode_2.11:0.1.3
      ...
    org.scalameta:contrib_2.11:1.6.0
      org.scalameta:scalameta_2.11:1.6.0
        ...
    org.scalameta:scalahost_2.11.8:1.6.0
      ...
    ...
```

`scalahost` brings in some of the same class files as `sourcecode` and `scalameta`. This is not a problem when running everything inside of a single, flat classloader like `java -cp` or `sbt` do, because one of each of the duplicates wins by being earlier on the flat classpath.

But it is a problem for non-private dependencies when tools like CBT or OSGI isolate classloaders for security and performance reasons. With non-private I mean when project has types of their dependencies in their public API. In scalafix sourcecode is probably private, but scalameta is public and brought in by both scalahost and transitively by contrib, which leads to link-time exceptions like `IncompatibleClassChange`. (For projects CBT optionally allows using a flat classloader, too, but not for builds at this time and scalafix is generally used as part of the build).

Publishing a version of scalahost that depends on those other libraries instead of including their class files should fix the problem. This was also discussed here: https://gitter.im/scala/tooling-contrib?at=58d7ec158e4b63533d66068d (and above). There may still have to be a flat version published until sbt fixes this: https://github.com/sbt/sbt/issues/2255

P.S. CBT is about to add a `verifyClasspath` task that finds such conflicts. Replacing contrib's scalameta dependency with scalahost works as a workaround for now.

**Comment by olafurpg** [comment]
Thank you for reporting! I was aware of this issue and had been planning to fix it for a while. I just opened a PR https://github.com/scalameta/scalameta/pull/775 to split scalahost into a `scalahost` module (published as regular library, no fat jar) and `scalahost-nsc` (compiler plugin, fat jar). 

Scalafix will depend on the `scalahost` module in the next release, avoiding classpath conflicts.

PS. Note, this change makes the `scala.meta.Mirror.apply` constructor require a path to the `scalafix-nsc` plugin fatjar. This is necessary to support runtime compilation of quasiquotes for symbol extraction (e.g., `q"scala.Predef".symbol"`).

**Comment by olafurpg** [comment]
It's possible to try out the new scalahost library module (no fatjar) with
```
resolvers += Resolver.bintrayIvyRepo("scalameta", "maven")
libraryDependencies += "org.scalameta" %% "scalameta" % "1.7.0-487-13087aaf"
libraryDependencies += "org.scalameta" %% "scalahost" % "1.7.0-487-13087aaf" cross CrossVersion.full
```

**Comment by cvogt** [comment]
Why does the scala meta mirror requires a scalafix fat jar? Would a scalahost fatjar be sufficient?

**Comment by cvogt** [comment]
> It's possible to try out the new scalahost library module (no fatjar) with

Awesome, thx :)!

**Comment by olafurpg** [comment]
It requires the `scalahost-nsc` fat jar for the same reason as here https://github.com/scalameta/scalameta/blob/13087aafb7e9db4d8926740fb86578bdd494ad6b/scalahost/core/src/main/scala/scala/meta/Mirror.scala#L25-L26

**Comment by cvogt** [comment]
Ah ok, I suppose you meant the scalahost-nsc plugin not the scalafix-nsc plugin in your previous comment then?

> this change makes the scala.meta.Mirror.apply constructor require a path to the scalafix-nsc plugin fatjar

**Comment by olafurpg** [comment]
Aha, yes, I meant scalahost-nsc ;)

**Comment by olafurpg** [comment]
Fixed as of 0.3.4

**Comment by cvogt** [comment]
nice :)!!
